src/index.js

```diff
@@ -86,6 +86,12 @@ function fetch (uri, opts) {
           reject(new FetchError(`redirect location header missing at: ${request.url}`, 'invalid-redirect'))
           return
         }
+        // Remove authorization if changing hostnames (but not if just
+        // changing ports or protocols).  This matches the behavior of request:
+        // https://github.com/request/request/blob/b12a6245/lib/redirect.js#L134-L138
+        if (url.parse(request.url).hostname !== url.parse(res.headers.location).hostname) {
```

Comment:
I think I solved the issues with parsing different types of urls and also added a couple more test cases. I wasn't sure about URIs like `foo.com` cause it seems those are not valid but maybe I'm misreading.  Lemme know if there is anymore changes I should make.